[jayden.v] 방탈출 예약 서비스 step3 제출합니다.

Tarakyu opened the pull request:
### 구현사항
- 테마 테이블 분리
- 테마에 대한 CRUD 구현
- 예약되어있는 테마는 삭제가 불가능하도록 검증

방탈출 예약 서비스 step3 제출합니다! 연관된 두 개의 테이블을 다루는 것을 중심으로 고민했습니다.
예약되어있는 테마는 삭제가 불가능하도록 검증하는 로직을 테스트하는 코드가 적절한 방법으로 설계되었는지도 궁금합니다.
감사합니다!

kakao-jayden-v committed:
chore: schema 변경

kakao-jayden-v committed:
feat: Theme Repository 구현

- Reservation 스키마 변경
- Reservation Mapping 구현 변경
- Theme Repository 2가지 구현
- ConnectionHandler 분리

kakao-jayden-v committed:
feat: Theme Service 구현

- Theme Service 구현
- Theme Repository update 메서드가 id를 매개변수로 받도록 수정

kakao-jayden-v committed:
fix: 기능 오류 수정

- Controller Advice 구문 오류 수정
- 콘솔 애플리케이션이 테마 스키마 추가에 대응할 수 있도록 수정
- 테마 업데이트 SQL문 문법 오류 수정
- 예약 생성 요청시 테마 ID를 입력받도록 수정

kakao-jayden-v committed:
feat: Theme Controller 구현

kakao-jayden-v committed:
feat: 이미 예약되어있는 테마는 삭제할 수 없도록 검증

kakao-jayden-v committed:
test: 이미 예약되어있는 테마는 삭제할 수 없도록 검증하는 로직의 테스트 추가

gogyzzz commented on src/main/java/nextstep/repository/ReservationH2JdbcTemplateRepository.java:
```diff
@@ -27,16 +27,14 @@ public ReservationH2JdbcTemplateRepository(JdbcTemplate jdbcTemplate) {
 
     @Override
     public Reservation add(Reservation reservation) {
```
여기서는 Exception 핸들링을 하지 않는 것 같은데요(delete도 마찬가지), 
잘못된 포맷으로 바디를 전달해보니 500 에러가 납니다. 
ControllerAdvice에서 별도 처리를 해줘야 할 것 같아요!

그리고 input validation 과정도 필요해보입니다~

또 add 라는 이름 대신 save가 어떨까요? JpaRepository에서는 save가 디폴트이름이라서요~

Tarakyu replied:
예약, 테마의 생성과 수정에 대해 예외처리 추가하였고, `add` 메서드의 이름도 `save`로 수정하였습니다!
[refactor: add 메서드의 이름을 save로 변경](https://github.com/next-step/spring-roomesacpe-reservation-kakao/pull/86/commits/e718dded2658d2005d1a3328ddbffbbb324587f7)

gogyzzz commented on src/main/java/nextstep/repository/ConnectionHandler.java:
```diff
@@ -0,0 +1,31 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public abstract class ConnectionHandler {
```
ConnectionHandler 가 JdbcTemplateRepository에서는 쓰이고 있지 않은 것 같아요~
(Spring이 아닌) RoomEscapeApplication 으로 만든 레코드는 RoomEscapeSpringApplication 에선 조회가 안되더라구요. 
그리고 RoomEscapeSpringApplication 으로 만든 레코드는 
임시(?)인지 어플리케이션이 종료되고 나면 다 날아가는 것 같은데 맞나요?

RoomEscapeSpringApplication 용으로 ConnectionHandler를 만든다면 println 대신 logger 를 사용해서 로깅을 하는 것도 필요해보입니다~

Tarakyu replied:
강의에서는 웹 애플리케이션은 인메모리 DB를 활용했었는데, 같은 DB를 활용하도록 개선해보았습니다.
`ConnectionHandler`는 웹 애플리케이션에서는 접근하지 않고 콘솔 애플리케이션에서만 접근하고 있습니다!
[feat: 웹 애플리케이션도 로컬 DB에 연결](https://github.com/next-step/spring-roomesacpe-reservation-kakao/pull/86/commits/250b87a23c0efebe9dcd0b6e664101b043331b96)

gogyzzz commented on src/main/java/nextstep/RoomEscapeApplication.java:
```diff
@@ -20,9 +22,11 @@ public class RoomEscapeApplication {
 
     public static void main(String[] args) {
```
main 의 내용이 너무 긴 것 같습니다~
indent 기준으로 봐도 많은 것 같아요~

https://edu.nextstep.camp/s/GiKTqpMP/ls/r1QiH4qK 하고 다른 신입분들 PR 보고 든 생각인데요, 
step3 부터는 예약 뿐만 아니라 테마 생성 조회 삭제 등의 작업도 ConsoleApplication에서 지원해야 되는 것 같습니다 (Spring 뿐만 아니라)

Tarakyu replied:
`main` 코드는 디폴트로 작성되어있던 코드를 그대로 두었었는데 이것도 한 번 리팩터링 해보았습니다~!!
그리고 콘솔에서도 테마 관리가 가능하도록 개선하였습니다.
[refactor: 콘솔 애플리케이션 리팩터링](https://github.com/next-step/spring-roomesacpe-reservation-kakao/pull/86/commits/72622fb4922442b9ef6ed28e99e75170c4dbf308)

gogyzzz commented on src/main/java/nextstep/service/ReservationService.java:
```diff
@@ -5,21 +5,25 @@
 import org.springframework.stereotype.Service;
 
 @Service
 public class ReservationService {
```
Reservation, ThemeService 둘 다 @Transactional 을 추가하는게 어떨까요?
메서드 도중에 예외가 발생하면 롤백이 되게끔 하는게 좋겠어요

Tarakyu replied:
그렇군요!! `@Transactional`에 대해서 잘 모르고 있었네요. 더 공부해보겠습니다 감사합니다 😋

gogyzzz commented on src/main/java/nextstep/repository/ReservationH2Repository.java:
```diff
@@ -42,10 +43,10 @@ public Reservation add(Reservation reservation) {
     @Override
     public Reservation findById(Long id) throws ReservationNotFoundException {
```
findById 의 리턴 타입을 Optional[Reservation]으로 바꾸는게 어떨까요? 
JpaRepository 를 쓰게 되면 findBy~ 메서드의 리턴타입이 Optional이 되거든요
그리고 예외는 Service나 Repository를 사용하는 쪽에서 던지는 것으로 처리하면 될 것 같아요

받는 쪽에선 받은 Optional 객체에 .orElseThrow 메서드를 사용해서 원하는 예외를 던지게 할 수 있습니다

Tarakyu replied:
말씀해주신대로 `Optional`을 반환하도록 수정하였습니다!

gogyzzz commented on src/main/java/nextstep/controller/RoomEscapeControllerAdvice.java:
```diff
@@ -23,4 +25,18 @@ public ResponseEntity<String> handleReservationNotFoundException(ReservationNotF
                 .status(HttpStatus.BAD_REQUEST)
                 .body(exception.getMessage());
     }
+
```
테마 등록 시 실패 예외 처리도 여기서 해야 하지 않을까 싶습니다
그리고 handle 메서드가 너무 많아지면 하나로만 사용하는 것도 방법일 것 같아요 
(대부분은 BAD_REQUEST에 메세지가 같이 나가는 형태니까요)

Tarakyu replied:
테마 등록 실패 예외 처리를 추가하였고, handle 메서드를 하나도 통일하였습니다!
[feat: 예외처리 보완](https://github.com/next-step/spring-roomesacpe-reservation-kakao/pull/86/commits/b4767110cc79861584d36fa7a2901c0a0a2b86e0)

kakao-jayden-v committed:
refactor: add 메서드의 이름을 save로 변경

kakao-jayden-v committed:
feat: 웹 애플리케이션도 로컬 DB에 연결

kakao-jayden-v committed:
feat: 예외처리 보완

- service 레이어에 Transactional 어노테이션 추가
- repository의 find 메서드가 Optional을 반환하도록 하고, service에서 예외처리하도록 변경
- ExceptionHandler에서 같은 status의 에러를 반환하는 메서드들을 하나로 병합
- input validation 추가
- 삭제 및 업데이트 요청 시 data를 찾을 수 없다면 에러 반환

kakao-jayden-v committed:
refactor: 콘솔 애플리케이션 리팩터링

- InputHandler 분리
- 테마 관리 기능 추가
- 테마 업데이트 구문 오류 수정

Tarakyu commented:
@gogyzzz 자세하게 리뷰해주셔서 감사합니다!! 프로젝트를 여러 방면에서 개선해나가면서 많이 공부가 되었어요. 🥺

gogyzzz commented:
지난주에 코멘트 드린 내용을 저도 다시 복기하느라 시간이 좀 걸렸네요 ㅋㅋ 
수정하시느라 고생하셨어요~
